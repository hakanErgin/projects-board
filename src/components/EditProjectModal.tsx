import React, { useState } from 'react';
import { useQuery, useMutation, gql } from '@apollo/client';
import { Modal, Button, Input, Select } from 'antd';

const { Option } = Select;

const GET_ENTERPRISES = gql`
  query {
    allEnterprises {
      id
      name
    }
  }
`;

const GET_PROJECT = gql`
  query($id: ID!) {
    Project(id: $id) {
      id
      name
      Enterprise {
        id
        name
      }
    }
  }
`;

const UPDATE_PROJECT = gql`
  mutation($id: ID!, $name: String, $enterprise_id: ID) {
    updateProject(id: $id, name: $name, enterprise_id: $enterprise_id) {
      id
      name
      enterprise_id
      Enterprise {
        id
        name
      }
    }
  }
`;

export function EditProjectModal(props: any) {
  const {
    selectedProjectId,
    isEditProjectModalVisible,
    setIsEditProjectModalVisible,
  } = props;

  const [selectedProjectName, setSelectedProjectName] = useState('');
  const [selectedEnterpriseId, setSelectedEnterpriseId] = useState('');

  const { loading, error, data } = useQuery(GET_ENTERPRISES);
  const { error: getProError } = useQuery(GET_PROJECT, {
    variables: { id: selectedProjectId },
    onCompleted: (data: any) => {
      setSelectedProjectName(data.Project.name);
      setSelectedEnterpriseId(data.Project.Enterprise.id);
    },
  });
  const [updateProjecct] = useMutation(UPDATE_PROJECT);

  if (loading) return <p>Loading...</p>;
  if (error || getProError) {
    return <p>Error</p>;
  }

  const { allEnterprises } = data;

  function onChange(value: any) {
    setSelectedEnterpriseId(value);
  }

  function handleOk() {
    updateProjecct({
      variables: {
        id: selectedProjectId,
        name: selectedProjectName,
        enterprise_id: selectedEnterpriseId,
      },
    });
    setIsEditProjectModalVisible(false);
  }

  function handleCancel(/* e: any */) {
    setIsEditProjectModalVisible(false);
  }

  if (selectedProjectName !== '') {
    return (
      <div className="EditProjectModal">
        <Modal
          title={`Edit ${selectedProjectName}`}
          visible={isEditProjectModalVisible}
          onOk={handleOk}
          onCancel={handleCancel}
          footer={
            <Button key="submit" type="primary" onClick={handleOk}>
              Edit Project
            </Button>
          }
        >
          <Input
            defaultValue={selectedProjectName}
            onChange={(e: any) => setSelectedProjectName(e.target.value)}
          />
          <Select
            // defaultValue={projectData.Project.Enterprise.Id}
            value={selectedEnterpriseId}
            showSearch
            style={{ width: '100%' }}
            // placeholder={allEnterprises[0].name}
            optionFilterProp="children"
            onChange={onChange}
            filterOption={(input, option: any) =>
              option.children.toLowerCase().indexOf(input.toLowerCase()) >= 0
            }
          >
            {allEnterprises.map((enterprise: any) => (
              <Option value={enterprise.id} key={enterprise.id}>
                {enterprise.name}
              </Option>
            ))}
          </Select>
        </Modal>
      </div>
    );
  } else return null;
}
