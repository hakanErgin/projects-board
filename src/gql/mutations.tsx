import { gql } from '@apollo/client';

export const UPDATE_PROJECT = gql`
  mutation($id: ID!, $name: String, $enterprise_id: ID) {
    updateProject(id: $id, name: $name, enterprise_id: $enterprise_id) {
      id
      name
      enterprise_id
      Enterprise {
        id
        name
      }
      Users {
        id
      }
    }
  }
`;

export const CREATE_PROJECT = gql`
  mutation($id: ID!, $name: String!, $enterprise_id: ID!) {
    createProject(id: $id, name: $name, enterprise_id: $enterprise_id) {
      id
    }
  }
`;

export const REMOVE_PROJECT = gql`
  mutation($id: ID!) {
    removeProject(id: $id)
  }
`;

export const UPDATE_USER = gql`
  mutation($id: ID!, $project_id: ID) {
    updateUser(id: $id, project_id: $project_id) {
      id
      project_id
    }
  }
`;
