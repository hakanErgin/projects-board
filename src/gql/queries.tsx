import { gql } from '@apollo/client';

export const GET_ENTERPRISES = gql`
  query {
    allEnterprises {
      id
      name
    }
  }
`;

export const GET_PROJECT = gql`
  query($id: ID!) {
    Project(id: $id) {
      id
      name
      Enterprise {
        id
        name
      }
    }
  }
`;

export const GET_PROJECTS = gql`
  query {
    allProjects {
      id
      name
      Users {
        id
      }
    }
  }
`;

export const GET_ALL_USERS = gql`
  query {
    allUsers {
      id
      email
      first_name
      last_name
    }
  }
`;

export const GET_USERS_BY_PROJECT = gql`
  query($id: ID!) {
    Project(id: $id) {
      Users {
        id
        first_name
        last_name
        avatar
        email
        project_id
      }
    }
  }
`;
