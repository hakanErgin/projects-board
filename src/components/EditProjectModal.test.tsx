import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { EditProjectModal } from './EditProjectModal';
import { MockedProvider } from '@apollo/client/testing';
import { GET_ENTERPRISES, GET_PROJECT } from '../gql';
import wait from 'waait';

// https://github.com/facebook/react/issues/11565
jest.mock('react-dom', () => {
  const original = jest.requireActual('react-dom');
  return {
    ...original,
    createPortal: (node: any) => node,
  };
});

const mockedQueries = [
  {
    request: {
      query: GET_ENTERPRISES,
    },
    result: {
      data: {
        allEnterprises: [
          {
            id: '1',
            name: 'Quop',
          },
          {
            id: '2',
            name: 'Blogspan',
          },
        ],
      },
    },
  },
  {
    request: {
      query: GET_PROJECT,
      variables: { id: '1' },
    },
    result: {
      data: {
        Project: {
          id: '1',
          name: 'Sub-Ex',
          Enterprise: {
            id: '1',
            name: 'Quop',
          },
        },
      },
    },
  },
];

it('renders correctly', async () => {
  const component = renderer.create(
    <MockedProvider mocks={mockedQueries} addTypename={false}>
      <EditProjectModal
        selectedProjectId="1"
        isEditProjectModalVisible={true}
        setIsEditProjectModalVisible={jest.fn()}
      />
    </MockedProvider>
  );
  await act(wait);
  expect(component).toMatchSnapshot();
});
