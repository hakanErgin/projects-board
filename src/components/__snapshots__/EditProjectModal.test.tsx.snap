// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="EditProjectModal"
>
  <div
    className="ant-modal-root"
  >
    <div
      className="ant-modal-mask"
      style={
        Object {
          "zIndex": undefined,
        }
      }
    />
    <div
      aria-labelledby="rcDialogTitle0"
      className="ant-modal-wrap"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="dialog"
      style={
        Object {
          "display": null,
          "zIndex": undefined,
        }
      }
      tabIndex={-1}
    >
      <div
        className="ant-modal Modal"
        onClick={[Function]}
        role="document"
        style={
          Object {
            "width": 520,
          }
        }
      >
        <div
          aria-hidden="true"
          style={
            Object {
              "height": 0,
              "outline": "none",
              "overflow": "hidden",
              "width": 0,
            }
          }
          tabIndex={0}
        />
        <div
          className="ant-modal-content"
        >
          <button
            aria-label="Close"
            className="ant-modal-close"
            onClick={[Function]}
            type="button"
          >
            <span
              className="ant-modal-close-x"
            >
              <span
                aria-label="close"
                className="anticon anticon-close ant-modal-close-icon"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="close"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                  />
                </svg>
              </span>
            </span>
          </button>
          <div
            className="ant-modal-header"
          >
            <div
              className="ant-modal-title"
              id="rcDialogTitle0"
            >
              Edit Sub-Ex
            </div>
          </div>
          <div
            className="ant-modal-body"
          >
            <div
              className="formContainer"
            >
              <span
                className="ant-typography"
                id="projectTitle"
                style={
                  Object {
                    "WebkitLineClamp": null,
                  }
                }
              >
                <strong>
                  Project's name
                </strong>
              </span>
              <input
                className="ant-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                type="text"
                value="Sub-Ex"
              />
            </div>
            <div
              className="formContainer"
            >
              <span
                className="ant-typography"
                id="enterpriseTitle"
                style={
                  Object {
                    "WebkitLineClamp": null,
                  }
                }
              >
                <strong>
                  Enterprise
                </strong>
              </span>
              <div
                className="ant-select ant-select-single ant-select-show-arrow ant-select-show-search"
                onBlur={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
              >
                <div
                  className="ant-select-selector"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                >
                  <span
                    className="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autoComplete="off"
                      className="ant-select-selection-search-input"
                      id="rc_select_TEST_OR_SSR"
                      onChange={[Function]}
                      onCompositionEnd={[Function]}
                      onCompositionStart={[Function]}
                      onKeyDown={[Function]}
                      onMouseDown={[Function]}
                      onPaste={[Function]}
                      readOnly={false}
                      role="combobox"
                      style={
                        Object {
                          "opacity": null,
                        }
                      }
                      type="search"
                      unselectable={null}
                      value=""
                    />
                  </span>
                  <span
                    className="ant-select-selection-item"
                    title="Quop"
                  >
                    Quop
                  </span>
                </div>
                <span
                  aria-hidden={true}
                  className="ant-select-arrow"
                  onMouseDown={[Function]}
                  style={
                    Object {
                      "WebkitUserSelect": "none",
                      "userSelect": "none",
                    }
                  }
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    className="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      className=""
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            className="ant-modal-footer"
          >
            <button
              className="ant-btn ant-btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <span>
                Edit Project
              </span>
            </button>
          </div>
        </div>
        <div
          aria-hidden="true"
          style={
            Object {
              "height": 0,
              "outline": "none",
              "overflow": "hidden",
              "width": 0,
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
  </div>
</div>
`;
